nolintlint: "allow no explanation" broken since v1.31.0

<details><summary>Version of golangci-lint</summary>

```console
$ golangci-lint --version
golangci-lint has version 1.33.0 built from b90551c on 2020-11-23T05:15:36Z
```

</details>

<details><summary>Config file</summary>

```console
$ cat .golangci.yml
run:
  timeout: "10m"
linters:
  disable-all: true
  enable:
    - "asciicheck"
    - "bodyclose"
    - "deadcode"
    - "depguard"
    - "errcheck"
    - "exportloopref"
    # - "gocognit"
    - "gocritic"
    - "gocyclo"
    - "godot"
    - "goerr113"
    - "gofmt"
    - "gofumpt"
    - "goimports"
    - "golint"
    - "govet"
    - "ineffassign"
    - "megacheck"
    - "misspell"
    - "nakedret"
    # - "nestif"
    - "nolintlint"
    - "structcheck"
    # - "stylecheck"
    - "unconvert"
    - "unparam"
    - "varcheck"
linters-settings:
  depguard:
    list-type: blacklist
    include-go-root: true
    packages-with-error-message:
      - errors: "use github.com/xxx/golang-libraries/errors"
      - reflect: "shouldn't be used by most application"
      - unsafe: "it's not safe"
      - github.com/pkg/errors: "use github.com/xxx/golang-libraries/errors"
      - golang.org/x/net/context: "use context"
      - golang.org/x/sync/singleflight: "use github.com/xxx/golang-libraries/singleflight"
  errcheck:
    check-type-assertions: true
  gocritic:
    enabled-tags:
      - diagnostic
      - style
      - performance
      - opinionated
      - experimental
    disabled-checks:
      - commentFormatting # TODO enable in order to follow the Go convention.
      - hugeParam # Too many problem for now. And maybe it's not a real issue.
      - paramTypeCombine # Too many false positive.
      - ptrToRefParam # TODO evaluate if this check can be enabled, and disable with nolint for specific cases.
      - rangeValCopy # TODO evaluate if this should be fixed or not.
      - unnamedResult # TODO fix projects.
      - whyNoLint # It's the developer/reviewer responsibility.
  gocyclo:
    min-complexity: 10
  govet:
    enable-all: true
    settings:
      printf:
        funcs:
          - "(github.com/xxx/golang-libraries/errors).Newf"
          - "(github.com/xxx/golang-libraries/errors).WithMessagef"
          - "(github.com/xxx/golang-libraries/errors).Wrapf"
  nolintlint:
    allow-unused: false
    allow-leading-space: false
    allow-no-explanation:
      - errcheck
      - misspell
    require-explanation: true
    require-specific: true
issues:
  exclude-use-default: false
  max-issues-per-linter: 0
  max-same-issues: 0
```

</details>

<details><summary>Go environment</summary>

```console
$ go version && go env
go version go1.16beta1 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/pierre/.cache/go-build"
GOENV="/home/pierre/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/pierre/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/pierre/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/pierre/.gimme/versions/go1.16beta1.src"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/pierre/.gimme/versions/go1.16beta1.src/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.16beta1"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/pierre/gosrc/github.com/xxx/golang-libraries/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2834066088=/tmp/go-build -gno-record-gcc-switches"
```

</details>

<details><summary>Verbose output of running</summary>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/pierre/gosrc/github.com/xxx/golang-libraries /home/pierre/gosrc/github.com/xxx /home/pierre/gosrc/github.com /home/pierre/gosrc /home/pierre /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 26 linters: [asciicheck bodyclose deadcode depguard errcheck exportloopref gocritic gocyclo godot goerr113 gofmt gofumpt goimports golint gosimple govet ineffassign misspell nakedret nolintlint staticcheck structcheck unconvert unparam unused varcheck] 
INFO [loader] Go packages loading at mode 575 (types_sizes|compiled_files|exports_file|files|imports|name|deps) took 336.576226ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 22.169636ms 
INFO [linters context/goanalysis] analyzers took 1m11.186014705s with top 10 stages: buildir: 18.196055998s, buildssa: 5.653807045s, godot: 3.979265239s, unconvert: 2.767148051s, goimports: 2.207601655s, golint: 1.950709086s, unparam: 1.913974376s, gocritic: 1.875576756s, inspect: 1.846274477s, gofumpt: 1.832033599s 
INFO [linters context/goanalysis] analyzers took 4.729446694s with top 10 stages: buildir: 4.073888816s, U1000: 655.557878ms 
INFO [runner] Issues before processing: 85, after processing: 0 
INFO [runner] Processors filtering stat (out/in): cgo: 85/85, skip_files: 85/85, skip_dirs: 85/85, identifier_marker: 85/85, path_prettifier: 85/85, exclude-rules: 85/85, exclude: 85/85, filename_unadjuster: 85/85, autogenerated_exclude: 85/85, nolint: 0/85 
INFO [runner] processing took 14.906659ms with stages: nolint: 12.62204ms, identifier_marker: 1.016951ms, path_prettifier: 633.51µs, autogenerated_exclude: 459.141µs, skip_dirs: 157.489µs, cgo: 9.877µs, filename_unadjuster: 3.974µs, uniq_by_line: 905ns, max_same_issues: 672ns, diff: 257ns, skip_files: 251ns, source_code: 229ns, max_from_linter: 225ns, severity-rules: 224ns, exclude: 194ns, sort_results: 164ns, exclude-rules: 162ns, max_per_file_from_linter: 153ns, path_shortener: 150ns, path_prefixer: 91ns 
INFO [runner] linters took 4.724833786s with stages: goanalysis_metalinter: 4.116987836s, unused: 592.85491ms 
INFO File cache stats: 367 entries of total size 750.1KiB 
INFO Memory: 52 samples, avg is 484.1MB, max is 748.3MB 
INFO Execution took 5.088920461s
```

</details>

The "allow no explanation" feature of `nolintlint` is broken since v1.31.0.
With v1.30.0 it's reporting "nolint" with no explanation.
With v1.31.0, it's not reporting anymore.

> I don't see any changes on nolintlint between v1.30.0 and v1.31.0:
Yes, that's why I'm reporting this issue

Here is a code sample:
```go
package foo

import "sync"

func foo() { //nolint:unused,deadcode
	var mu sync.Mutex
	mu.Lock()
	mu.Unlock() //nolint:staticcheck
}
```

With golangci-lint v1.30.0:
```
/home/pierre/go/pkg/golangci-lint/v1.30.0/golangci-lint -v run
INFO [config_reader] Config search paths: [./ /home/pierre/gosrc/github.com/xxx/golang-libraries /home/pierre/gosrc/github.com/xxx /home/pierre/gosrc/github.com /home/pierre/gosrc /home/pierre /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 26 linters: [asciicheck bodyclose deadcode depguard errcheck exportloopref gocritic gocyclo godot goerr113 gofmt gofumpt goimports golint gosimple govet ineffassign misspell nakedret nolintlint staticcheck structcheck unconvert unparam unused varcheck] 
INFO [loader] Go packages loading at mode 575 (deps|types_sizes|compiled_files|exports_file|files|imports|name) took 339.61536ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 28.848484ms 
INFO [linters context/goanalysis] analyzers took 55.070695101s with top 10 stages: buildir: 20.120179537s, buildssa: 6.73736716s, unconvert: 2.2801639s, gocritic: 1.732114317s, inspect: 1.591059712s, fact_deprecated: 1.564588276s, goimports: 1.435186059s, gofumpt: 1.431137375s, unparam: 1.093742404s, findcall: 815.953598ms 
INFO [linters context/goanalysis] analyzers took 4.790057642s with top 10 stages: buildir: 4.134919134s, U1000: 655.138508ms 
INFO [runner] Issues before processing: 185, after processing: 5 
INFO [runner] Processors filtering stat (out/in): filename_unadjuster: 185/185, path_prettifier: 185/185, skip_dirs: 185/185, max_same_issues: 5/5, cgo: 185/185, autogenerated_exclude: 185/185, identifier_marker: 185/185, nolint: 5/185, uniq_by_line: 5/5, severity-rules: 5/5, skip_files: 185/185, exclude-rules: 185/185, diff: 5/5, source_code: 5/5, exclude: 185/185, max_per_file_from_linter: 5/5, max_from_linter: 5/5, path_shortener: 5/5, path_prefixer: 5/5, sort_results: 5/5 
INFO [runner] processing took 18.857354ms with stages: nolint: 11.774195ms, identifier_marker: 4.243026ms, path_prettifier: 1.233804ms, autogenerated_exclude: 1.025312ms, skip_dirs: 425.223µs, cgo: 74.976µs, filename_unadjuster: 46.004µs, source_code: 26.916µs, uniq_by_line: 2.715µs, path_shortener: 1.862µs, max_same_issues: 968ns, max_per_file_from_linter: 698ns, exclude: 315ns, skip_files: 290ns, diff: 260ns, severity-rules: 178ns, sort_results: 177ns, max_from_linter: 173ns, exclude-rules: 169ns, path_prefixer: 93ns 
INFO [runner] linters took 4.982791242s with stages: goanalysis_metalinter: 4.337452436s, unused: 626.391297ms 
foo/foo.go:5:14: directive `//nolint:unused,deadcode` should provide explanation such as `//nolint:unused,deadcode // this is why` (nolintlint)
func foo() { //nolint:unused,deadcode
             ^
foo/foo.go:8:14: directive `//nolint:staticcheck` should provide explanation such as `//nolint:staticcheck // this is why` (nolintlint)
	mu.Unlock() //nolint:staticcheck
	            ^
INFO File cache stats: 370 entries of total size 752.6KiB 
INFO Memory: 55 samples, avg is 479.2MB, max is 745.6MB 
INFO Execution took 5.355913258s
```

With golangci-lint v1.31.0:
```
/home/pierre/go/pkg/golangci-lint/v1.31.0/golangci-lint -v run
INFO [config_reader] Config search paths: [./ /home/pierre/gosrc/github.com/xxx/golang-libraries /home/pierre/gosrc/github.com/xxx /home/pierre/gosrc/github.com /home/pierre/gosrc /home/pierre /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 26 linters: [asciicheck bodyclose deadcode depguard errcheck exportloopref gocritic gocyclo godot goerr113 gofmt gofumpt goimports golint gosimple govet ineffassign misspell nakedret nolintlint staticcheck structcheck unconvert unparam unused varcheck] 
INFO [loader] Go packages loading at mode 575 (files|deps|exports_file|imports|name|types_sizes|compiled_files) took 336.069307ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 10.685933ms 
INFO [linters context/goanalysis] analyzers took 1m38.268075127s with top 10 stages: buildir: 20.327705409s, buildssa: 7.189830674s, unconvert: 2.708765339s, gocritic: 2.117544207s, gofmt: 2.116390408s, gofumpt: 1.899475029s, golint: 1.816105463s, fact_deprecated: 1.669929366s, inspect: 1.669397434s, misspell: 1.472912402s 
INFO [linters context/goanalysis] analyzers took 5.10636342s with top 10 stages: buildir: 4.287906332s, U1000: 818.457088ms 
INFO [runner] Issues before processing: 90, after processing: 0 
INFO [runner] Processors filtering stat (out/in): filename_unadjuster: 90/90, skip_files: 90/90, skip_dirs: 90/90, cgo: 90/90, exclude-rules: 90/90, nolint: 0/90, identifier_marker: 90/90, path_prettifier: 90/90, autogenerated_exclude: 90/90, exclude: 90/90 
INFO [runner] processing took 15.241814ms with stages: nolint: 12.836075ms, identifier_marker: 1.016901ms, path_prettifier: 605.119µs, autogenerated_exclude: 597.325µs, skip_dirs: 165.807µs, cgo: 11.412µs, filename_unadjuster: 4.808µs, max_same_issues: 1.101µs, uniq_by_line: 607ns, diff: 432ns, source_code: 317ns, max_from_linter: 302ns, skip_files: 259ns, sort_results: 256ns, path_shortener: 235ns, severity-rules: 210ns, exclude-rules: 199ns, max_per_file_from_linter: 179ns, exclude: 179ns, path_prefixer: 91ns 
INFO [runner] linters took 5.29990061s with stages: goanalysis_metalinter: 4.611727345s, unused: 672.396532ms 
INFO File cache stats: 368 entries of total size 750.9KiB 
INFO Memory: 58 samples, avg is 479.9MB, max is 749.2MB 
INFO Execution took 5.651591992s
```

Same config file as the one I shared above.

--- a/pkg/golinters/nolintlint/nolintlint.go
+++ b/pkg/golinters/nolintlint/nolintlint.go
@@ -145,12 +145,10 @@ func (l Linter) Run(fset *token.FileSet, nodes ...ast.Node) ([]Issue, error) {
 	for _, node := range nodes {
 		if file, ok := node.(*ast.File); ok {
 			for _, c := range file.Comments {
-				text := c.Text()
-				matches := directiveOnlyPattern.FindStringSubmatch(text)
-				if len(matches) == 0 {
+				directive := getText(c)
+				if directive == "" {
 					continue
 				}
-				directive := matches[1]
 
 				// check for a space between the "//" and the directive
 				leadingSpaceMatches := leadingSpacePattern.FindStringSubmatch(c.List[0].Text) // c.Text() doesn't have all leading space
@@ -236,4 +234,18 @@ func (l Linter) Run(fset *token.FileSet, nodes ...ast.Node) ([]Issue, error) {
 		}
 	}
 	return issues, nil
+}
+
+func getText(c *ast.CommentGroup) string {
+	var text string
+	for _, comment := range c.List {
+		if !directiveOnlyPattern.MatchString(comment.Text) {
+			continue
+		}
+
+		text = strings.TrimSpace(strings.TrimPrefix(comment.Text, "//"))
+		break
+	}
+
+	return text
 }

--- a/pkg/golinters/nolintlint/nolintlint.go
+++ b/pkg/golinters/nolintlint/nolintlint.go
@@ -167,14 +167,12 @@ func (l Linter) Run(fset *token.FileSet, nodes ...ast.Node) ([]Issue, error) {
 					base := BaseIssue{
 						fullDirective:                     comment.Text,
 						directiveWithOptionalLeadingSpace: directiveWithOptionalLeadingSpace,
-						position:                          fset.Position(c.Pos()),
+						position:                          fset.Position(comment.Pos()),
 					}
 
 					// check for, report and eliminate leading spaces so we can check for other issues
 					if len(leadingSpace) > 1 {
-						issues = append(issues, ExtraLeadingSpace{
-							BaseIssue: base,
-						})
+						issues = append(issues, ExtraLeadingSpace{BaseIssue: base})
 					}
 
 					if (l.needs&NeedsMachineOnly) != 0 && len(leadingSpace) > 0 {

--- a/pkg/golinters/nolintlint/nolintlint.go
+++ b/pkg/golinters/nolintlint/nolintlint.go
@@ -47,7 +47,7 @@ type NotSpecific struct {
 }
 
 func (i NotSpecific) Details() string {
-	return fmt.Sprintf("directive `%s` should mention specific linter such as `//%s:my-linter`",
+	return fmt.Sprintf("directive `%s` should mention specific linter such as `%s:my-linter`",
 		i.fullDirective, i.directiveWithOptionalLeadingSpace)
 }
 
@@ -58,7 +58,7 @@ type ParseError struct {
 }
 
 func (i ParseError) Details() string {
-	return fmt.Sprintf("directive `%s` should match `//%s[:<comma-separated-linters>] [// <explanation>]`",
+	return fmt.Sprintf("directive `%s` should match `%s[:<comma-separated-linters>] [// <explanation>]`",
 		i.fullDirective,
 		i.directiveWithOptionalLeadingSpace)
 }
@@ -112,8 +112,7 @@ const (
 	NeedsAll = NeedsMachineOnly | NeedsSpecific | NeedsExplanation
 )
 
-// matches lines starting with the nolint directive
-var directiveOnlyPattern = regexp.MustCompile(`^\s*(nolint)(:\s*[\w-]+\s*(?:,\s*[\w-]+\s*)*)?\b`)
+var commentPattern = regexp.MustCompile(`^//\s*(nolint)(:\s*[\w-]+\s*(?:,\s*[\w-]+\s*)*)?\b`)
 
 // matches a complete nolint directive
 var fullDirectivePattern = regexp.MustCompile(`^//\s*nolint(:\s*[\w-]+\s*(?:,\s*[\w-]+\s*)*)?\s*(//.*)?\s*\n?$`)
@@ -142,110 +141,104 @@ var trailingBlankExplanation = regexp.MustCompile(`\s*(//\s*)?$`)
 
 func (l Linter) Run(fset *token.FileSet, nodes ...ast.Node) ([]Issue, error) {
 	var issues []Issue
+
 	for _, node := range nodes {
 		if file, ok := node.(*ast.File); ok {
 			for _, c := range file.Comments {
-				directive := getText(c)
-				if directive == "" {
-					continue
-				}
-
-				// check for a space between the "//" and the directive
-				leadingSpaceMatches := leadingSpacePattern.FindStringSubmatch(c.List[0].Text) // c.Text() doesn't have all leading space
-				if len(leadingSpaceMatches) == 0 {
-					continue
-				}
-				leadingSpace := leadingSpaceMatches[1]
-
-				directiveWithOptionalLeadingSpace := directive
-				if len(leadingSpace) > 0 {
-					directiveWithOptionalLeadingSpace = " " + directive
-				}
-
-				base := BaseIssue{
-					fullDirective:                     c.List[0].Text,
-					directiveWithOptionalLeadingSpace: directiveWithOptionalLeadingSpace,
-					position:                          fset.Position(c.Pos()),
-				}
-
-				// check for, report and eliminate leading spaces so we can check for other issues
-				if leadingSpace != "" && leadingSpace != " " {
-					issues = append(issues, ExtraLeadingSpace{
-						BaseIssue: base,
-					})
-				}
-
-				if (l.needs&NeedsMachineOnly) != 0 && strings.HasPrefix(directiveWithOptionalLeadingSpace, " ") {
-					issues = append(issues, NotMachine{BaseIssue: base})
-				}
-
-				fullMatches := fullDirectivePattern.FindStringSubmatch(c.List[0].Text)
-				if len(fullMatches) == 0 {
-					issues = append(issues, ParseError{BaseIssue: base})
-					continue
-				}
-				lintersText, explanation := fullMatches[1], fullMatches[2]
-				var linters []string
-				if len(lintersText) > 0 {
-					lls := strings.Split(lintersText[1:], ",")
-					linters = make([]string, 0, len(lls))
-					for _, ll := range lls {
-						ll = strings.TrimSpace(ll)
-						if ll != "" {
-							linters = append(linters, ll)
-						}
+				for _, comment := range c.List {
+					if !commentPattern.MatchString(comment.Text) {
+						continue
 					}
-				}
-				if (l.needs & NeedsSpecific) != 0 {
-					if len(linters) == 0 {
-						issues = append(issues, NotSpecific{BaseIssue: base})
-					}
-				}
 
-				// when detecting unused directives, we send all the directives through and filter them out in the nolint processor
-				if l.needs&NeedsUnused != 0 {
-					if len(linters) == 0 {
-						issues = append(issues, UnusedCandidate{BaseIssue: base})
-					} else {
-						for _, linter := range linters {
-							issues = append(issues, UnusedCandidate{BaseIssue: base, ExpectedLinter: linter})
-						}
-					}
-				}
+					// check for a space between the "//" and the directive
+					leadingSpaceMatches := leadingSpacePattern.FindStringSubmatch(comment.Text)
 
-				if (l.needs&NeedsExplanation) != 0 && (explanation == "" || strings.TrimSpace(explanation) == "//") {
-					needsExplanation := len(linters) == 0 // if no linters are mentioned, we must have explanation
-					// otherwise, check if we are excluding all of the mentioned linters
-					for _, ll := range linters {
-						if !l.excludeByLinter[ll] { // if a linter does require explanation
-							needsExplanation = true
-							break
-						}
+					var leadingSpace string
+					if len(leadingSpaceMatches) > 0 {
+						leadingSpace = leadingSpaceMatches[1]
 					}
-					if needsExplanation {
-						fullDirectiveWithoutExplanation := trailingBlankExplanation.ReplaceAllString(c.List[0].Text, "")
-						issues = append(issues, NoExplanation{
-							BaseIssue:                       base,
-							fullDirectiveWithoutExplanation: fullDirectiveWithoutExplanation,
+
+					directiveWithOptionalLeadingSpace := comment.Text
+					if len(leadingSpace) > 0 {
+						split := strings.Split(strings.SplitN(comment.Text, ":", 2)[0], "//")
+						directiveWithOptionalLeadingSpace = "// " + strings.TrimSpace(split[1])
+					}
+
+					base := BaseIssue{
+						fullDirective:                     comment.Text,
+						directiveWithOptionalLeadingSpace: directiveWithOptionalLeadingSpace,
+						position:                          fset.Position(c.Pos()),
+					}
+
+					// check for, report and eliminate leading spaces so we can check for other issues
+					if len(leadingSpace) > 1 {
+						issues = append(issues, ExtraLeadingSpace{
+							BaseIssue: base,
 						})
+					}
+
+					if (l.needs&NeedsMachineOnly) != 0 && len(leadingSpace) > 0 {
+						issues = append(issues, NotMachine{BaseIssue: base})
+					}
+
+					fullMatches := fullDirectivePattern.FindStringSubmatch(comment.Text)
+					if len(fullMatches) == 0 {
+						issues = append(issues, ParseError{BaseIssue: base})
+						continue
+					}
+
+					lintersText, explanation := fullMatches[1], fullMatches[2]
+					var linters []string
+					if len(lintersText) > 0 {
+						lls := strings.Split(lintersText[1:], ",")
+						linters = make([]string, 0, len(lls))
+						for _, ll := range lls {
+							ll = strings.TrimSpace(ll)
+							if ll != "" {
+								linters = append(linters, ll)
+							}
+						}
+					}
+
+					if (l.needs & NeedsSpecific) != 0 {
+						if len(linters) == 0 {
+							issues = append(issues, NotSpecific{BaseIssue: base})
+						}
+					}
+
+					// when detecting unused directives, we send all the directives through and filter them out in the nolint processor
+					if (l.needs & NeedsUnused) != 0 {
+						if len(linters) == 0 {
+							issues = append(issues, UnusedCandidate{BaseIssue: base})
+						} else {
+							for _, linter := range linters {
+								issues = append(issues, UnusedCandidate{BaseIssue: base, ExpectedLinter: linter})
+							}
+						}
+					}
+
+					if (l.needs&NeedsExplanation) != 0 && (explanation == "" || strings.TrimSpace(explanation) == "//") {
+						needsExplanation := len(linters) == 0 // if no linters are mentioned, we must have explanation
+						// otherwise, check if we are excluding all of the mentioned linters
+						for _, ll := range linters {
+							if !l.excludeByLinter[ll] { // if a linter does require explanation
+								needsExplanation = true
+								break
+							}
+						}
+
+						if needsExplanation {
+							fullDirectiveWithoutExplanation := trailingBlankExplanation.ReplaceAllString(comment.Text, "")
+							issues = append(issues, NoExplanation{
+								BaseIssue:                       base,
+								fullDirectiveWithoutExplanation: fullDirectiveWithoutExplanation,
+							})
+						}
 					}
 				}
 			}
 		}
 	}
+
 	return issues, nil
-}
-
-func getText(c *ast.CommentGroup) string {
-	var text string
-	for _, comment := range c.List {
-		if !directiveOnlyPattern.MatchString(comment.Text) {
-			continue
-		}
-
-		text = strings.TrimSpace(strings.TrimPrefix(comment.Text, "//"))
-		break
-	}
-
-	return text
 }

--- a/pkg/golinters/nolintlint/nolintlint_test.go
+++ b/pkg/golinters/nolintlint/nolintlint_test.go
@@ -41,6 +41,20 @@ func foo() {
 			},
 		},
 		{
+			desc:  "when multiple directives on multiple lines",
+			needs: NeedsExplanation,
+			contents: `
+package bar
+
+// example
+//nolint // this is ok
+//nolint:dupl
+func foo() {}`,
+			expected: []string{
+				"directive `//nolint:dupl` should provide explanation such as `//nolint:dupl // this is why` at testing.go:6:1",
+			},
+		},
+		{
 			desc:     "when no explanation is needed for a specific linter",
 			needs:    NeedsExplanation,
 			excludes: []string{"lll"},

--- a/pkg/golinters/nolintlint/nolintlint_test.go
+++ b/pkg/golinters/nolintlint/nolintlint_test.go
@@ -6,12 +6,22 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+//nolint:funlen
 func TestNoLintLint(t *testing.T) {
-	t.Run("when no explanation is provided", func(t *testing.T) {
-		linter, _ := NewLinter(NeedsExplanation, nil)
-		expectIssues(t, linter, `
+	testCases := []struct {
+		desc     string
+		needs    Needs
+		excludes []string
+		contents string
+		expected []string
+	}{
+		{
+			desc:  "when no explanation is provided",
+			needs: NeedsExplanation,
+			contents: `
 package bar
 
 func foo() {
@@ -21,25 +31,27 @@ func foo() {
   good() //nolint // this is ok
 	other() //nolintother
 }`,
-			"directive `//nolint` should provide explanation such as `//nolint // this is why` at testing.go:5:9",
-			"directive `//nolint //` should provide explanation such as `//nolint // this is why` at testing.go:6:9",
-			"directive `//nolint // ` should provide explanation such as `//nolint // this is why` at testing.go:7:9",
-		)
-	})
-
-	t.Run("when no explanation is needed for a specific linter", func(t *testing.T) {
-		linter, _ := NewLinter(NeedsExplanation, []string{"lll"})
-		expectIssues(t, linter, `
+			expected: []string{
+				"directive `//nolint` should provide explanation such as `//nolint // this is why` at testing.go:5:9",
+				"directive `//nolint //` should provide explanation such as `//nolint // this is why` at testing.go:6:9",
+				"directive `//nolint // ` should provide explanation such as `//nolint // this is why` at testing.go:7:9",
+			},
+		},
+		{
+			desc:     "when no explanation is needed for a specific linter",
+			needs:    NeedsExplanation,
+			excludes: []string{"lll"},
+			contents: `
 package bar
 
 func foo() {
 	thisIsAReallyLongLine() //nolint:lll
-}`)
-	})
-
-	t.Run("when no specific linter is mentioned", func(t *testing.T) {
-		linter, _ := NewLinter(NeedsSpecific, nil)
-		expectIssues(t, linter, `
+}`,
+		},
+		{
+			desc:  "when no specific linter is mentioned",
+			needs: NeedsSpecific,
+			contents: `
 package bar
 
 func foo() {
@@ -47,35 +59,41 @@ func foo() {
   bad() //nolint
   bad() // nolint // because
 }`,
-			"directive `//nolint` should mention specific linter such as `//nolint:my-linter` at testing.go:6:9",
-			"directive `// nolint // because` should mention specific linter such as `// nolint:my-linter` at testing.go:7:9")
-	})
-
-	t.Run("when machine-readable style isn't used", func(t *testing.T) {
-		linter, _ := NewLinter(NeedsMachineOnly, nil)
-		expectIssues(t, linter, `
+			expected: []string{
+				"directive `//nolint` should mention specific linter such as `//nolint:my-linter` at testing.go:6:9",
+				"directive `// nolint // because` should mention specific linter such as `// nolint:my-linter` at testing.go:7:9",
+			},
+		},
+		{
+			desc:  "when machine-readable style isn't used",
+			needs: NeedsMachineOnly,
+			contents: `
 package bar
 
 func foo() {
   bad() // nolint
   good() //nolint
-}`, "directive `// nolint` should be written without leading space as `//nolint` at testing.go:5:9")
-	})
-
-	t.Run("extra spaces in front of directive are reported", func(t *testing.T) {
-		linter, _ := NewLinter(0, nil)
-		expectIssues(t, linter, `
+}`,
+			expected: []string{
+				"directive `// nolint` should be written without leading space as `//nolint` at testing.go:5:9",
+			},
+		},
+		{
+			desc: "extra spaces in front of directive are reported",
+			contents: `
 package bar
 
 func foo() {
   bad() //  nolint
   good() // nolint
-}`, "directive `//  nolint` should not have more than one leading space at testing.go:5:9")
-	})
-
-	t.Run("spaces are allowed in comma-separated list of linters", func(t *testing.T) {
-		linter, _ := NewLinter(0, nil)
-		expectIssues(t, linter, `
+}`,
+			expected: []string{
+				"directive `//  nolint` should not have more than one leading space at testing.go:5:9",
+			},
+		},
+		{
+			desc: "spaces are allowed in comma-separated list of linters",
+			contents: `
 package bar
 
 func foo() {
@@ -84,40 +102,42 @@ func foo() {
   good() // nolint: linter1,linter2
   good() // nolint: linter1, linter2
 }`,
-			"directive `// nolint:linter1 linter2` should match `// nolint[:<comma-separated-linters>] [// <explanation>]` at testing.go:6:9", //nolint:lll // this is a string
-		)
-	})
-
-	t.Run("multi-line comments don't confuse parser", func(t *testing.T) {
-		linter, _ := NewLinter(0, nil)
-		expectIssues(t, linter, `
+			expected: []string{
+				"directive `// nolint:linter1 linter2` should match `// nolint[:<comma-separated-linters>] [// <explanation>]` at testing.go:6:9", //nolint:lll // this is a string
+			},
+		},
+		{
+			desc: "multi-line comments don't confuse parser",
+			contents: `
 package bar
 
 func foo() {
   //nolint:test
   // something else
-}`)
-	})
-}
+}`,
+		},
+	}
 
-func expectIssues(t *testing.T, linter *Linter, contents string, issues ...string) {
-	actualIssues := parseFile(t, linter, contents)
-	actualIssueStrs := make([]string, 0, len(actualIssues))
-	for _, i := range actualIssues {
-		actualIssueStrs = append(actualIssueStrs, i.String())
-	}
-	assert.ElementsMatch(t, issues, actualIssueStrs, "expected %s but got %s", issues, actualIssues)
-}
+	for _, test := range testCases {
+		test := test
+		t.Run(test.desc, func(t *testing.T) {
+			t.Parallel()
 
-func parseFile(t *testing.T, linter *Linter, contents string) []Issue {
-	fset := token.NewFileSet()
-	expr, err := parser.ParseFile(fset, "testing.go", contents, parser.ParseComments)
-	if err != nil {
-		t.Fatalf("unable to parse file contents: %s", err)
+			linter, _ := NewLinter(test.needs, test.excludes)
+
+			fset := token.NewFileSet()
+			expr, err := parser.ParseFile(fset, "testing.go", test.contents, parser.ParseComments)
+			require.NoError(t, err)
+
+			actualIssues, err := linter.Run(fset, expr)
+			require.NoError(t, err)
+
+			actualIssueStrs := make([]string, 0, len(actualIssues))
+			for _, i := range actualIssues {
+				actualIssueStrs = append(actualIssueStrs, i.String())
+			}
+
+			assert.ElementsMatch(t, test.expected, actualIssueStrs, "expected %s \nbut got %s", test.expected, actualIssues)
+		})
 	}
-	issues, err := linter.Run(fset, expr)
-	if err != nil {
-		t.Fatalf("unable to parse file: %s", err)
-	}
-	return issues
 }

--- a/pkg/golinters/nolintlint/nolintlint_test.go
+++ b/pkg/golinters/nolintlint/nolintlint_test.go
@@ -24,6 +24,8 @@ func TestNoLintLint(t *testing.T) {
 			contents: `
 package bar
 
+// example
+//nolint
 func foo() {
   bad() //nolint
   bad() //nolint //
@@ -32,9 +34,10 @@ func foo() {
 	other() //nolintother
 }`,
 			expected: []string{
-				"directive `//nolint` should provide explanation such as `//nolint // this is why` at testing.go:5:9",
-				"directive `//nolint //` should provide explanation such as `//nolint // this is why` at testing.go:6:9",
-				"directive `//nolint // ` should provide explanation such as `//nolint // this is why` at testing.go:7:9",
+				"directive `//nolint` should provide explanation such as `//nolint // this is why` at testing.go:5:1",
+				"directive `//nolint` should provide explanation such as `//nolint // this is why` at testing.go:7:9",
+				"directive `//nolint //` should provide explanation such as `//nolint // this is why` at testing.go:8:9",
+				"directive `//nolint // ` should provide explanation such as `//nolint // this is why` at testing.go:9:9",
 			},
 		},
 		{

--- a/test/errchk.go
+++ b/test/errchk.go
@@ -22,7 +22,7 @@ import (
 //
 // Sources files are supplied as fullshort slice.
 // It consists of pairs: full path to source file and its base name.
-//nolint:gocyclo,funlen
+//nolint:gocyclo
 func errorCheck(outStr string, wantAuto bool, fullshort ...string) (err error) {
 	var errs []error
 	out := splitOutput(outStr, wantAuto)
@@ -160,7 +160,7 @@ var (
 )
 
 // wantedErrors parses expected errors from comments in a file.
-//nolint:nakedret,gocyclo,funlen
+//nolint:nakedret
 func wantedErrors(file, short string) (errs []wantedError) {
 	cache := make(map[string]*regexp.Regexp)
 

